contribute/design/1.x/replica_scaleup/dataplane/20190910-replica-scaleup.md

```diff
@@ -0,0 +1,278 @@
+
+### High Level Design
+cstor-volume-mgmt watches for CStorVolume CR and updates istgt.conf file if
+there is any change in DesiredReplicationFactor.
+It will trigger istgtcontrol command so that istgt updates DRF during runtime.
+
+A listener will be created on UnixDomain socket in cstor-volume-mgmt container.
+istgt connects to this listener to update the replica related information if
+there is any change in the list.
+cstor-volume-mgmt updates the status.ReplicaList part of CStorVolume CR and
+sends the success/failure of CR update as reponse to istgt.
+
+During start phase of cstor-volume-mgmt container, it updates the istgt.conf
+file with values from spec.DesiredReplicationFactor and status.ReplicaList of
+CStorVolume CR.
+
+#### Components interaction:
+
+```
+[Conf file] --->  ISTGT <--- REPLICAS
+                  |  /|\
+   (UnixSocket Conn)  |
+                  |   | 
+                  |  (istgtcontrol) 
+                 \|/  | 
+            CSTOR-VOLUME-MGMT <---->[CStorVolume CR]                          V
+                    | 
+                    | 
+                   \|/ 
+               [Conf file]
+```
+
+#### Sample YAMLs
+
+#### NOTES
+
+##### Is there a need to maintain the list of known replicas:
+This can be better explained with scenarios.
+1. Consider case where 3 replicas (R1, R2, R3) are connected and are fine.
+Lets say, one replica(R3) was lost and new one (R4) got added.
+After R4 got reconstructed with data, another replica (R2) got lost and
+new one (R5) got connected.
+Once R5 got reconstructed, all replicas have quorum propery on. But, just that
+R2 and R3 are not reachable.
+At a later point of time, if R1, R4, R5 are down and R2, R3 are connected to
+istgt (which can be due to attachment of old disks), istgt cannot identify that
+there is missing data with R2, R3.
+
+2. Consider replacing non-existing replica.
+When user reduced replication factor, and later increased it, if old replicas
+gets connected, istgt can't identify that there is missing data with replicas.
+
+### Low Level Design
+Current code takes care of reconstructing data to non-quorum replica once it
+does handshake with target. But, changes are required in allowing replica to
+perform handshake with target to achieve DesiredReplicationFactor.
+
+#### Replica Connection
+Below are the steps to allow replica handshake with istgt:
+- If currently connected replicas count >= 5, reject this if it is non-quorum or
+another connected non-quorum replica if the new one is in quorum
+- If DesiredRF number of replicas are healthy, reject
+- Make sure connected replica is an appropriate one to the volume, by checking
+replica name with vol name
+- If quorum is off, allow replica to connect and follow existing code flow
+(i.e, adding to non_quorum linked list etc)
+- If quorum is on, allow to connect only if it is new replica i.e.,
+its checkpointed nums are zero.
+- If quorum is on and its checkpointed nums are not zero, allow to connect only
```

Comment:
May be a section can be added below how to handle cases where older replicas are trying to connect, while the volume is healthy with new replicas. It appears, the option is to delete the CVR.

Reply:
yes @kmova .. you are right, option is to delete them. I'm trying to avoid it thinking of some mistakes that users can make. We can do this (i.e., allowing older replicas to reconnect) once we build operators around it. I will mention this explicitly as non-goal.